README.md
MojoX-Mysql
===========

Mojolicious ♥ Mysql


Changes
0.23  2016-05-20
  - Trace SQL

0.22  2016-05-12
  - debug mode $ENV{'MOJO_MYSQL_DEBUG'} = 1

0.21  2016-03-17
  - fixed logging sql

0.19  2016-02-28
  - added logging sql

0.18  2016-02-26
  - new migration
  - restore code

0.14  2015-05-19
  - added fake data command

0.13  2015-05-19
  - added migration command

0.12  2015-02-15
  - Fixed build

0.11  2015-02-15
  - Fixed bug with empty time

0.09  2015-01-08
  - Small fixed type int 0

0.08  2015-01-07
  - Fixed type int 0

0.07  2014-12-29
  - Added automatic conversion in the field "datetime" to Mojo::Date

0.06  2014-11-04
  - Skip test

0.05  2014-11-04
  - Beta version



t/migration/fake.pm
package migration::fake;

1;

__DATA__

@@ 1
INSERT INTO `test1` (`text`) VALUES('Тест');



t/migration/default.pm
package migration::default;

1;

__DATA__

@@ 1
CREATE TABLE IF NOT EXISTS `test1` (
	`test_id` int unsigned NOT NULL AUTO_INCREMENT,
	`text` varchar(65534) NOT NULL,
	PRIMARY KEY (`test_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

@@ 2
CREATE TABLE IF NOT EXISTS `test2` (
	`test_id` int unsigned NOT NULL AUTO_INCREMENT,
	`text` varchar(65534) NOT NULL,
	PRIMARY KEY (`test_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;




lib/MojoX/Mysql/DB.pm
package MojoX::Mysql::DB;
use Mojo::Base -base;
use Mojo::Util qw(dumper);
use Carp qw(croak);

has [qw(id)] => '_default';

sub DESTROY {
	my $self = shift;
	$self->rollback;
	$self->disconnect;
	$self->flush;
}

# Master connect DB
sub connect_master {
	my $self = shift;
	my $id   = $self->id;
	$self->flush;

	my $config = {};
	if(exists $self->{'config'}->{$id}->{'master'}){
		$config = $self->{'config'}->{$id}->{'master'};
	}
	else{
		croak 'Not found id server';
	}

	if(defined $config && ref($config) eq 'HASH' && %{$config}){
		if(ref($self->{'dbh'}{'master'}{$id}) eq 'DBI::db'){
			my $dbh = $self->{'dbh'}{'master'}{$id};
			return $dbh if($dbh->ping);
		}

		my $dbh = DBI->connect("DBI:mysql:".$config->{'dsn'}, $config->{'user'}, $config->{'password'}, {
			AutoCommit=>0,
			RaiseError=>0,
			PrintError=>0,
			mysql_enable_utf8=>1,
			mysql_auto_reconnect=>1,
			mysql_write_timeout=>$config->{'write_timeout'},
			mysql_read_timeout=>$config->{'read_timeout'},
		});

		if($DBI::errstr){
			$self->{'dbh'}{'master'}{$id} = undef;
			croak $DBI::errstr;
			return;
		}
		else{
			$self->{'dbh'}{'master'}{$id} = $dbh;
			$dbh->{'RaiseError'} = 1;
			return $dbh;
		}
	}
	return;
}

# Slave connect DB
sub connect_slave {
	my $self = shift;
	my $id   = $self->id;
	$self->flush;

	my $config = {};
	if(exists $self->{'config'}{$id}->{'slave'}){
		$config = $self->{'config'}->{$id}->{'slave'};
	}
	else{
		croak 'Not found id server';
	}

	if(defined $config && ref($config) eq 'ARRAY' && @{$config}){
		if(ref($self->{'dbh'}{'slave'}{$id}) eq 'DBI::db'){
			my $dbh = $self->{'dbh'}{'slave'}{$id};
			return $dbh if($dbh->ping);
		}

		for my $conf (@{$config}){

			my $dbh = DBI->connect("DBI:mysql:".$conf->{'dsn'}, $conf->{'user'}, $conf->{'password'}, {
				AutoCommit=>0,
				RaiseError=>0,
				PrintError=>0,
				mysql_enable_utf8=>1,
				mysql_auto_reconnect=>1,
				mysql_write_timeout=>$conf->{'write_timeout'},
				mysql_read_timeout=>$conf->{'read_timeout'},
			});

			if($DBI::errstr){
				$self->{'dbh'}{'slave'}{$id} = undef;
				next;
			}
			else{
				$self->{'dbh'}{'slave'}{$id} = $dbh;
				$dbh->{'RaiseError'} = 1;
				return $dbh;
			}
		}
	}
	return;
}

sub commit {
	my ($self) = @_;
	while(my($id,$types) = each %{$self->{'dbh'}}){
		if(ref $types eq 'HASH'){
			for my $type (keys %{$types}){
				my $dbh = $self->{'dbh'}->{$id}->{$type};
				$dbh->commit if(ref $dbh eq 'DBI::db');
				warn "commit:$id,$type" if(defined $ENV{'MOJO_MYSQL_DEBUG'});
			}
		}
	}
}

sub rollback {
	my ($self) = @_;
	while(my($id,$types) = each %{$self->{'dbh'}}){
		if(ref $types eq 'HASH'){
			for my $type (keys %{$types}){
				my $dbh = $self->{'dbh'}->{$id}->{$type};
				$dbh->rollback if(ref $dbh eq 'DBI::db');
				warn "rollback:$id,$type" if(defined $ENV{'MOJO_MYSQL_DEBUG'});
			}
		}
	}
}

sub disconnect {
	my ($self) = @_;
	while(my($id,$types) = each %{$self->{'dbh'}}){
		if(ref $types eq 'HASH'){
			for my $type (keys %{$types}){
				my $dbh = $self->{'dbh'}->{$id}->{$type};
				$dbh->disconnect if(ref $dbh eq 'DBI::db');
				warn "disconnect:$id,$type" if(defined $ENV{'MOJO_MYSQL_DEBUG'});
				delete $self->{'dbh'};
			}
		}
	}
}

sub flush {
	my $self = shift;
	$self->id('_default');
}

1;


=encoding utf8

=head1 NAME

MojoX::Mysql::DB - Connect DB
 
=head1 SYNOPSIS

    use MojoX::Mysql;
    use Mojo::Util qw(dumper);

    my %config = (
        user=>'root',
        password=>undef,
        server=>[
            {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', type=>'master'},
            {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', type=>'slave'},
            {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>1, type=>'master'},
            {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>1, type=>'slave'},
            {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>2, type=>'master'},
            {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>2, type=>'slave'},
        ]
    );

    my $mysql = MojoX::Mysql->new(%config);
    my $dbh = $mysql->db->connect_master;

=head1 ATTRIBUTES

=head2 id

    $mysql->id(1); # choice id server

=head1 METHODS

=head2 connect_master

    $mysql->db->connect_master;

Return connect object.

=head2 connect_slave

    $mysql->db->connect_slave;

Return connect object.

=head2 commit

    $mysql->db->commit;

Commit on all servers

=head2 rollback

    $mysql->db->rollback;

Rollback on all servers

=head2 disconnect

    $mysql->db->disconnect;

Disconnect on all servers

=cut


lib/MojoX/Mysql/Result.pm
package MojoX::Mysql::Result;
use Mojo::Base -base;
use Mojo::Util qw(dumper);
use Mojo::Collection 'c';
use Mojo::Date;

sub async {
	my ($self,$sth,$dbh,$cb) = @_;
	sleep(0.01) until($sth->mysql_async_ready);
	my $counter = $sth->mysql_async_result;
	my $collection = $self->collection($sth,$cb);
	$sth->finish;
	$dbh->commit;
	$dbh->disconnect;
	return wantarray ? ($collection,$counter,$sth,$dbh) : $collection;
}

sub collection {
	my ($self,$sth,$cb) = @_;
	my $collection = c();
	my $names = $sth->{'NAME'};
	my $types = $sth->{'mysql_type_name'};
	my $nulls = $sth->{'NULLABLE'};
	#my $scale = $sth->{'Statement'};

	while (my $ref = $sth->fetch()) {
		if(ref($names) eq 'ARRAY'){
			my %hash;
			my $count_state = -1;
			for(@{$names}){
				$count_state++;
				my $value = $ref->[$count_state];
				my $type = $types->[$count_state];
				my $null = $nulls->[$count_state];

				if($type eq 'tinyint' || $type eq 'smallint' || $type eq 'mediumint' || $type eq 'integer' || $type eq 'bigint'){
					if($null == 1 && !defined $value){
						$value = undef;
					}
					else{
						$value = int $value;
					}
				}
				elsif($type eq 'datetime' && defined $value){
					$value = Mojo::Date->new($value);
				}
				else{
					if(!$value && $null){
						$value = undef;
					}
					else{
						$value =~ s/^\s+|\s+$//g;
						utf8::decode($value) unless utf8::is_utf8($value);
					}
				}
				$hash{$_} = $value;
			}
			$self->$cb(\%hash) if(ref $cb eq 'CODE');
			push(@{$collection}, \%hash);
		}
	}
	return $collection;
}


1;



lib/Mojolicious/Plugin/Mysql.pm
package Mojolicious::Plugin::Mysql;
use Mojo::Base 'Mojolicious::Plugin';
use Mojo::Util qw(dumper);
use FindBin;
use lib "$FindBin::Bin/../lib/";
use MojoX::Mysql;

sub register {
	my ($self, $app, $config) = @_;

	my $mysql = MojoX::Mysql->new(%{$config});

	# MojoX
	$app->helper(mysql=>sub {
		my ($c) = @_;
		$mysql->app($c->app);
		return $mysql;
	});

	# Hook Commit
	$app->hook(after_dispatch => sub {
		my $self = shift;
		my $exception = $self->stash('exception');
		if(defined $exception){
			$self->mysql->db->rollback();
		}
		else{
			$self->mysql->db->commit();
			$self->mysql->db->disconnect();
		}
	});
	return $mysql;
}

1;

=encoding utf8

=head1 SYNOPSIS

    my %config = (
	    user=>'root',
	    password=>undef,
	    server=>[
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', type=>'master'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', type=>'slave'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>1, type=>'master'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>1, type=>'slave'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>2, type=>'master'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>2, type=>'slave'},
	    ]
    );

    # Mojolicious
    $self->plugin('Mysql'=>\%config);

    # Mojolicious::Lite
    plugin 'Mysql' => \%config;

=head1 HELPERS

=head2 mysql

    $app->mysql;

Return L<MojoX::Mysql> object.

=cut



lib/Mojolicious/Command/sql.pm
package Mojolicious::Command::sql;
use Mojo::Base 'Mojolicious::Command';
use FindBin;
use Getopt::Long qw(GetOptionsFromArray :config no_auto_abbrev no_ignore_case);
use Mojo::Util qw(dumper encode);
use Mojo::Date;
use Term::ANSIColor;
use Mojo::Loader qw(data_section load_class);

has description => encode 'UTF-8', "sql data migrations\n";
has usage => <<EOF;

    usage: $0 sql [create|delete|update]

EOF

sub run {
	my ($self, $action, $id) = @_;

	$id = '_default' if(!defined $id);
	my $migration_object = $self->app->mysql->{'migration'}->{$id};
	my $e = load_class $migration_object;
	warn qq{Loading "$migration_object" failed: $e} if ref $e;

	my $all = data_section($migration_object);
	my @table = ();
	my @version = ();
	while(my ($id,$sql) = each(%{$all})){
		$sql =~ s/\t//g;
		$sql =~ s/\n//g;
		$sql =~ s/`//g;
		if($sql =~ m/^CREATE\s+TABLE\sIF\sNOT\sEXISTS\s(?<table>([\w]+))/i || $sql =~ m/^CREATE\s+TABLE\s(?<table>([\w]+))/i){
			push(@table,$+{'table'});
		}
		push(@version,$id);
	}

	if(defined $action && $action eq 'delete'){
		$self->app->mysql->id($id)->do("SET FOREIGN_KEY_CHECKS = 0;");
		$self->app->mysql->id($id)->do("DROP TABLE IF EXISTS `$_`;") for (@table);
		$self->app->mysql->id($id)->do("DROP TABLE IF EXISTS `_version`;");
		$self->app->mysql->id($id)->do("SET FOREIGN_KEY_CHECKS = 1;");
		$self->app->mysql->db->commit;
	}
	elsif(defined $action && $action eq 'create'){
		$self->app->mysql->id($id)->do('CREATE TABLE IF NOT EXISTS `_version` (`version_id` int unsigned NOT NULL, `date` datetime NOT NULL, PRIMARY KEY (`version_id`)) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;');
		$self->app->mysql->id($id)->do('INSERT INTO `_version` (`version_id`,`date`) VALUES(0,UTC_TIMESTAMP());');
		$self->app->mysql->db->commit;
	}
	elsif(defined $action && $action eq 'update'){
		my $collection = $self->app->mysql->query('SELECT `version_id` FROM `_version` ORDER BY `version_id` DESC LIMIT 1;');
		my $version_id = $collection->last->{'version_id'};
		say colored ['bright_yellow'],'current sql version:'.$version_id;

		for my $file_version (sort {$a <=> $b} @version){
			if($file_version > int $version_id){
				my $sql = data_section($migration_object,$file_version);
				if(defined $sql){
					$sql =~ s/\t//g;
					$sql =~ s/\n//g;
					$self->app->mysql->id($id)->do($sql);
					$self->app->mysql->id($id)->do("INSERT INTO `_version` (`version_id`,`date`) VALUES($file_version,UTC_TIMESTAMP());");
					say colored ['bright_green'],'update sql version:'.$file_version;
				}
			}
			else {
				say colored ['bright_red'],'skip sql version:'.$file_version;
			}
		}
	}
	elsif(defined $action && $action eq 'fake' && $self->app->mode eq 'development'){
		my $fake_object = $self->app->mysql->{'fake'}->{$id};
		my $e = load_class $fake_object;
		warn qq{Loading "$fake_object" failed: $e} if ref $e;

		my @version = ();
		while(my ($id) = each(%{$all})){
			push(@version,$id);
		}

		for my $version (sort {$a <=> $b} @version){
			my $sql = data_section($fake_object,$version);
			if(defined $sql){
				$sql =~ s/\t//g;
				$sql =~ s/\n//g;
				$self->app->mysql->id($id)->do($sql);
				say colored ['bright_blue'],'fake sql version:'.$version;
			}
		}
	}
	else{
		say $self->usage;
	}
	$self->app->mysql->db->commit;
	return $self;
}

1;

__DATA__

=encoding utf8

=head1 SYNOPSIS

    my %config = (
	    user=>'root',
	    password=>undef,
	    server=>[
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', type=>'master', migration=>'migration::default'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', type=>'slave'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>1, type=>'master', migration=>'migration::default1'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>1, type=>'slave'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>2, type=>'master', migration=>'migration::default2'},
		    {dsn=>'database=test;host=localhost;port=3306;mysql_connect_timeout=5;', id=>2, type=>'slave'},
	    ]
    );

    Usage: APPLICATION sql [ACTION] [ID]

        ./myapp.pl sql create # first create
        ./myapp.pl sql update # update sql
        ./myapp.pl sql delete # delete all table

        ./myapp.pl sql update 1 # update sql id 1


=head1 MIGRATION PACKAGE

    package migration::default;

    1;

    __DATA__

    @@ 1
    CREATE TABLE IF NOT EXISTS `test1` (
	    `test_id` int unsigned NOT NULL AUTO_INCREMENT,
	    `text` varchar(65534) NOT NULL,
	    PRIMARY KEY (`test_id`)
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

    @@ 2
    CREATE TABLE IF NOT EXISTS `test2` (
	    `test_id` int unsigned NOT NULL AUTO_INCREMENT,
	    `text` varchar(65534) NOT NULL,
	    PRIMARY KEY (`test_id`)
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

=cut

